### PR TITLE
Fix S3878 AD0001: Cover the case of ImplicitArrayCreationExpressionSyntax

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs [hunk 1]
```diff
@@ -40,7 +40,12 @@ public sealed class ArrayPassedAsParams : ArrayPassedAsParamsBase<SyntaxKind, Ar
             : null;
 
     protected override ITypeSymbol ArrayElementType(ArgumentSyntax argument, SemanticModel model) =>
-        model.GetTypeInfo(((ArrayCreationExpressionSyntax)argument.Expression).Type.ElementType).Type;
+        argument switch
+        {
+            _ when argument.Expression as ArrayCreationExpressionSyntax is { } arrayCreation => model.GetTypeInfo(arrayCreation.Type.ElementType).Type,
+            _ when argument.Expression as ImplicitArrayCreationExpressionSyntax is { } implicitArrayCreation => model.GetTypeInfo(implicitArrayCreation).Type,
+            _ => null
+        };
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs:48]:
```suggestion
        argument.Expression switch
        {
            ArrayCreationExpressionSyntax arrayCreation => model.GetTypeInfo(arrayCreation.Type.ElementType).Type,
            ImplicitArrayCreationExpressionSyntax implicitArrayCreation => model.GetTypeInfo(implicitArrayCreation).Type,
            _ => null
        };
```

**Review by Tim-Pohlmann — APPROVED** [review]
The switch can be simplified, looks good otherwise.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs [hunk 2]
```diff
@@ -40,7 +40,12 @@ public sealed class ArrayPassedAsParams : ArrayPassedAsParamsBase<SyntaxKind, Ar
             : null;
 
     protected override ITypeSymbol ArrayElementType(ArgumentSyntax argument, SemanticModel model) =>
-        model.GetTypeInfo(((ArrayCreationExpressionSyntax)argument.Expression).Type.ElementType).Type;
+        argument switch
+        {
+            _ when argument.Expression as ArrayCreationExpressionSyntax is { } arrayCreation => model.GetTypeInfo(arrayCreation.Type.ElementType).Type,
+            _ when argument.Expression as ImplicitArrayCreationExpressionSyntax is { } implicitArrayCreation => model.GetTypeInfo(implicitArrayCreation).Type,
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: This looks wrong. It does return the array type, but the method is called ArrayElementType. It should be `model.GetTypeInfo(implicitArrayCreation).Type as IArrayTypeSymbol)?.ElementType`

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: good catch!

Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: maybe more UTs are required then.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: Done

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ArrayPassedAsParams.cs [hunk 3]
```diff
@@ -81,25 +81,27 @@ public class Repro6894
...
+        MethodMixed(1, new String[] { "1", "2" });                              // Noncompliant
+        MethodArray(new String[] { "1", "2" }, new String[] { "1", "2" });      // Compliant, elements in args: [System.String[], System.String[]]
+        MethodArray(new int[] { 1, 2 }, new int[] { 1, 2 });                    // Compliant, elements in args: [System.Int32[], System.Int32[]]
+        MethodJaggedArray(new int[] { 1, 2 });                                  // Compliant: jagged array [System.Object[]]
+        Method(d);                                                              // Compliant
+        Method("Hello", 2);                                                     // Compliant
         string.Format(CultureInfo.InvariantCulture, "{0}.{1}", new object[] { "", new object() });
+        MethodImplicitArray(new[] { "Hello", "Hi" });                           // Noncompliant , Implicit array creation
+        Method(new [] { 1, 2, 3, });                                            // Compliant, Elements in args: [System.Int32[]]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ArrayPassedAsParams.cs:104]:
```suggestion
        Method(new[] { 1, 2, 3, });                                            // Compliant, Elements in args: [System.Int32[]]
```

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/ArrayPassedAsParams.cs:104]: Add one more test with a multidimensional array
```suggestion
        Method(new[] { 1, 2, 3, });                                            // Compliant, Elements in args: [System.Int32[]]
        Method(new[,] { { 1, 2 } });                                            // Compliant, Elements in args: [System.Int32[,]]        
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: Do the UTs cover differences in the array rank? I don't think so, and maybe the implementation as well:
```cs
void Method(params object[] args) { }

Method(new object[] { null }); // Noncompliant
// These cases might be FPs
Method(new object[,] { { null } }); // Compliant
Method(new[,] { { new object() } }); // Compliant

// This should work
Method(new object[][] { new[] { new object() } }); // Compliant
```

params array must be single-dimensional.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs]: We cover only jagged arrays.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ArrayPassedAsParams.cs [hunk 4]
```diff
@@ -81,25 +81,32 @@ public class Repro6894
...
+        MethodArray(new String[] { "1", "2" }, new String[] { "1", "2" });      // Compliant, elements in args: [System.String[], System.String[]]
+        MethodArray(new int[] { 1, 2 }, new int[] { 1, 2 });                    // Compliant, elements in args: [System.Int32[], System.Int32[]]
+        MethodJaggedArray(new int[] { 1, 2 });                                  // Compliant: jagged array [System.Object[]]
+        Method(d);                                                              // Compliant
+        Method("Hello", 2);                                                     // Compliant
         string.Format(CultureInfo.InvariantCulture, "{0}.{1}", new object[] { "", new object() });
+        MethodImplicitArray(new[] { "Hello", "Hi" });                           // Noncompliant , Implicit array creation
+        Method(new[] { 1, 2, 3, });                                             // Compliant, Elements in args: [System.Int32[]]
+        Method(new[,] { { 1, 2 } });                                            // Compliant, Elements in args: [System.Int32[,]]
+        Method(new object[] { null });                                          // Compliant
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ArrayPassedAsParams.cs:106]: Why is `Method(new String[] { "1", "2" }); ` Noncompliant, while `Method(new object[] { null });` is compliant? This looks like an FN to me.

**Review by martin-strecker-sonarsource — APPROVED** [review]
Approved, because the cases left are edge cases, but there seems to be something wrong with the implementation. I think there is some special handling with `object[]` params and I'm not sure, we handle multi-dimensional arrays properly.
You may want to have another look later.

**Pull request merged**